proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,260 @@
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+To help maintain the total capital balance and track/verify the reward
+distribution during rewarding phases, a new SysVar account, `EpochRewards`, is
+proposed.
+
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks for distribution in the next phase.
+
+Currently, on Solana Mainnet Beta with ~550K active stake accounts, it shows
+that epoch reward calculation takes around 10 seconds on average. This will
+make it impossible to perform rewards computation synchronous within one block.
+To make this work, a design of complete async reward computation is required.
+
+However, there are quite a few promising optimizations that can cut down the
+reward computation time. An experiment for reward calculation optimization
+(https://github.com/solana-labs/solana/pull/31193) showed that we can cut
+the reward calculation time down to **40ms**. This is well below the 400ms
+block time.
+
+Therefore, the following of the design is based on the above optimization. The
+reward calculation will be performed at the first block of the epoch. The full
+rewards result is calculated, the result is partitioned into distribution
+chunks stored in the bank, which will then be distributed during the `reward
+distribution` phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+First, the reward results are sorted by account's Pubkey, and randomly shuffled
+with a fast `rng` seeded by current epoch. The shuffled reward results are then
+divided into `M` chunks. This process will ensure that the reward distribution
+is deterministic for the current epoch, while also introducing a degree of
+randomness between epochs.
+
+To minimize the impact on block processing time during the reward distribution
+phase, a total of 4,096 accounts will be distributed per block. The total
+number of blocks needed to distributed rewards is given by the following
+formula to avoid using floating point value arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+### `EpochRewards` Sysvar Account
+
+`EpochRewards` sysvar account records the total rewards plus some other reward
+related information internally. And the account balance reflects the amount of
+pending rewards to distribute.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+
+```
+struct RewardRewards{
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
```

Comment:
I remember that @t-nelson mentioned that it may be possible for some other accounts to transfer Lamport to the sysvar accounts. Then, total_reward_in_lamport - sysvar balance < distributed_reward_in_lamport. 

IMO, distributed_reward_in_lamport is a good redundancy to have.

Reply:
Ah, yep. wfm